cpp/src/arrow/compute/kernels/scalar_string_test.cc

```diff
@@ -2119,6 +2119,96 @@ TYPED_TEST(TestStringKernels, SliceCodeunitsNegPos) {
 
 #endif  // ARROW_WITH_UTF8PROC
 
+TYPED_TEST(TestBaseBinaryKernels, SliceBytesBasic) {
+  SliceOptions options{2, 4};
+  this->CheckUnary("binary_slice_bytes", R"(["foo", "fo", null, "foo "])", this->type(),
+                   R"(["o", "", null, "o "])", &options);
+
+  // end is beyond 0, but before start (hence empty)
+  SliceOptions options_edgecase_1{-3, 1};
+  this->CheckUnary("binary_slice_bytes", R"(["foods"])", this->type(), R"([""])",
+                   &options_edgecase_1);
+
+  // this is a safeguard agains an optimization path possible, but actually a tricky case
+  SliceOptions options_edgecase_2{-6, -2};
+  this->CheckUnary("binary_slice_bytes", R"(["foods"])", this->type(), R"(["foo"])",
+                   &options_edgecase_2);
+
+  auto input = ArrayFromJSON(this->type(), R"(["foods"])");
+  EXPECT_RAISES_WITH_MESSAGE_THAT(
+      Invalid,
+      testing::HasSubstr(
+          "Function 'binary_slice_bytes' cannot be called without options"),
+      CallFunction("binary_slice_bytes", {input}));
+
+  SliceOptions options_invalid{2, 4, 0};
+  EXPECT_RAISES_WITH_MESSAGE_THAT(
+      Invalid, testing::HasSubstr("Slice step cannot be zero"),
+      CallFunction("binary_slice_bytes", {input}, &options_invalid));
+}
+
+TYPED_TEST(TestBaseBinaryKernels, SliceBytesPosPos) {
+  SliceOptions options{2, 4};
+  this->CheckUnary("binary_slice_bytes", R"(["", "f", "fo", "foo", "food", "foods"])",
+                   this->type(), R"(["", "", "", "o", "od", "od"])", &options);
+  SliceOptions options_step{1, 5, 2};
+  this->CheckUnary("binary_slice_bytes", R"(["", "f", "fo", "foo", "food", "foods"])",
+                   this->type(), R"(["", "", "o", "o", "od", "od"])", &options_step);
+  SliceOptions options_step_neg{5, 1, -2};
+  this->CheckUnary("binary_slice_bytes", R"(["", "f", "fo", "foo", "food", "foods"])",
+                   this->type(), R"(["", "", "", "o", "d", "so"])", &options_step_neg);
+  options_step_neg.stop = 0;
+  this->CheckUnary("binary_slice_bytes", R"(["", "f", "fo", "foo", "food","foods"])",
+                   this->type(), R"(["", "", "o", "o", "do", "so"])", &options_step_neg);
+}
+
+TYPED_TEST(TestBaseBinaryKernels, SliceBytesPosNeg) {
+  SliceOptions options{2, -1};
+  this->CheckUnary("binary_slice_bytes", R"(["", "f", "fo", "foo", "food", "foods"])",
```

Comment:
These tests would be better without any letter repetition in the source values. Otherwise the results might end up correct even with an incorrect implementation.

Reply:
Also, it would be nice to put some non-ASCII bytes in there as well, to check that slicing is really byte-wise.